include/zephyr/drivers/gpio.h

```diff
@@ -752,6 +754,96 @@ static inline int gpio_pin_configure_dt(const struct gpio_dt_spec *spec,
 				  spec->dt_flags | extra_flags);
 }
 
+/*
+ * @brief Get direction of all pins in a port.
+ *
+ * If @p inputs or @p outputs is NULL, then this function does not get the
+ * respective input or output direction information.
+ *
+ * @param port Pointer to the device structure for the driver instance.
+ * @param inputs Pointer to a variable where input directions will be stored.
+ * @param outputs Pointer to a variable where output directions will be stored.
+ *
+ * @retval 0 If successful.
+ * @retval -EIO I/O error when accessing an external GPIO chip.
+ * @retval -EWOULDBLOCK if operation would block.
+ */
+__syscall int gpio_port_get_direction(const struct device *port, gpio_port_pins_t *inputs,
+					   gpio_port_pins_t *outputs);
+
+static inline int z_impl_gpio_port_get_direction(const struct device *port,
+						      gpio_port_pins_t *inputs,
+						      gpio_port_pins_t *outputs)
+{
+	if (!IS_ENABLED(CONFIG_GPIO_GET_DIRECTION)) {
+		return -ENOTSUP;
+	}
```

Comment:
Do we need this as lines 784-786 has a similar check.

Reply:
Yeah, could probably remove it. I guess technically if each driver uses `IS_ENABLED()` in the driver API, then it's not needed.

Reply:
Apologies to bring this up again, but why do we need the Kconfig `GPIO_GET_DIRECTION`
Shouldn't the below check be able to address platforms that do not implement the new API to get the GPIO direction.
```
       if (api->port_get_direction_raw == NULL) {
               return -ENOTSUP;
       }
```